Migrate index instead of creating a new one if it exists

LucaGallinari opened the pull request:
While the populate command is running, you won't get any results for any queries made to ElasticSearch (taxon, search and instant). This is probably due to the fact the the new empty Index that is created is, somehow, immediatly seen as the "active" index, even if the command is still populating it.
I found out [this issue on Elastically](https://github.com/jolicode/elastically/issues/52) where they say this:

> As far as I understand, when I need to update my index while keeping the old data available, I need to call **migrate**, markAsLive and purgeOldIndices

delyriand commented on src/Model/Document/Index/Indexer.php:
```diff
@@ -101,7 +101,12 @@ public function indexAll(): void
     {
         $indexName = $this->getIndexName($locale);
-        $newIndex = $this->getIndexBuilder()->createIndex($indexName);
+        $index = $this->getClient()->getIndex($indexName);
```
the real problem is the `markAsLive` call before the flush. I offer another solution, can you try this:

```diff
    public function indexAllByLocale(string $locale): void
    {
        $indexName = $this->getIndexName($locale);
        $newIndex = $this->getIndexBuilder()->createIndex($indexName);
-       $this->getIndexBuilder()->markAsLive(
-           $newIndex,
-           $indexName
-       );

        $repositories = $this->documentRepositoryProvider->getRepositories();
        foreach ($repositories as $repository) {
            $documents = $repository->findAll();
            /** @var DocumentableInterface $document */
            foreach ($documents as $document) {
                Assert::isInstanceOf($document, DocumentableInterface::class);
                $this->indexOneByLocale($document->convertToDocument($locale), $locale);
            }
        }

        $this->getIndexer()->flush();

+       $this->getIndexBuilder()->markAsLive(
+           $newIndex,
+           $indexName
+       );
        $this->getIndexer()->refresh($indexName);
        try {
            $this->getIndexBuilder()->purgeOldIndices($indexName);
        } catch (ResponseException $exception) {
            throw new ReadOnlyIndexException($exception->getMessage());
        }
    }
```

With this version, the alias is added after the index is populated.

maximehuran replied:
WDYTOT @LucaGallinari

LucaGallinari replied:
Ok, i tried with this fix and while populate is running i still see the results in frontend, but when it ends there aren't any results and the newly created index is empty. Even with other populate it won't show any results.
And if i start from a clear DB (without indices) i get an error:

> In Bulk.php line 337:
>   Error in one or more bulk request actions:
>   index: /documents-it/_doc/product-1 caused mapper [attributes.value] of different type, current_type [boolean], merged_type [text]
>   index: /documents-it/_doc/product-2 caused mapper [attributes.value] of different type, current_type [boolean], merged_type [text]
>   index: /documents-it/_doc/product-3 caused mapper [attributes.value] of different type, current_type [boolean], merged_type [text]
> ...

delyriand replied:
It's an another problem… I think you have a checkbox type product attribute.  
Can you try adding typing to string in \MonsieurBiz\SyliusSearchPlugin\Model\Documentable\DocumentableProductTrait::addAttributesInDocument:

```diff
    protected function addAttributesInDocument(ResultInterface $document, string $locale): ResultInterface
    {
        /** @var AttributeValueInterface $attribute */
        foreach ($this->getAttributesByLocale($locale, $locale) as $attribute) {
            $attributeValues = [];
            if (isset($attribute->getAttribute()->getConfiguration()['choices'])) {
                foreach ($attribute->getValue() as $value) {
                    $attributeValues[] = $attribute->getAttribute()->getConfiguration()['choices'][$value][$locale];
                }
            } else {
-               $attributeValues[] = $attribute->getValue();
+              $attributeValues[] = (string) $attribute->getValue();
            }
            $document->addAttribute($attribute->getCode(), $attribute->getName(), $attributeValues, $attribute->getLocaleCode() ?? $locale, 1);
        }

        return $document;
    }
```

But other types of product attributes, like date and datetime, will also cause errors.

delyriand replied:
ping @LucaGallinari

LucaGallinari replied:
Ok, i tried now. Even with that it still reports the mapping problem. And btw i don't think that's "It's an another problem…" cause this only occurs when i put the markAsLive after the flush where you indicated.
And it's already 3 days that we are in production with the "migrate" and we haven't had any issues.

lruozzi9 commented:
**ATTENTION**!! There is a huge problem using migrate index. As specified in the issue reported by @LucaGallinari (https://github.com/jolicode/elastically/issues/52), the need for that issue was to "merge" results from two indexes. This could **NOT** be applied to Sylius.

We had documents that were not supposed to be indexed in the query results. After searching, I found that they had been indexed on ES in a transition phase of the DB (so the information that was used to tell him not to index that product was missing), but they were never deleted in the next indexes as the migrate function keeps always the old documents merging them with the new ones. This was our case, but the same thing could apply, for example, to a product that is no longer salable.

LucaGallinari commented:
As stated by Lorenzo in the previous comment, migrate is not the correct action to use in this case, so we reverted to the "markAsLive" one and put this call after the documents indexing as suggested by you in the comments. As i wrote this solution was not complete cause the new index was eventually empty. We digged in a little bit and found out that the problem is the call to indexOneByLocale: this method makes another call to getClient()->getIndex() instead of using the newly created index. So it was trying to put the documents in the wrong index (maybe the old one, that had these documents already). This led to the new index to be empty.
We removed the call to the indexOneByLocale method and made the same call that it does but with the new index. This is seems to work properly.

maximehuran commented:
Thank you @LucaGallinari and @lruozzi9 🚀
